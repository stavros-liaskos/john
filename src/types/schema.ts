/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {
  '/': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/artist/search': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          pattern?: string;
          size?: number;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['ArtistSearchResponse'];
          };
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/artists': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          page?: number;
          size?: number;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/artists/recommended': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          page?: number;
          size?: number;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['FollowedArtistsResponse'];
          };
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/me': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/me/enable-services': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          enableSpotify?: boolean;
          lastfmUsername?: string;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/me/follow': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get?: never;
    put?: never;
    post: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: {
        content: {
          'application/json': components['schemas']['SearchResultArtistDto'];
        };
      };
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/me/followed-artists': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['FollowedArtistsResponse'];
          };
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/me/unfollow/{artistId}': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get?: never;
    put?: never;
    post?: never;
    delete: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          artistId: number;
        };
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/notify-users': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'text/plain': boolean;
          };
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/raccoon-user': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          email?: string;
          id?: number;
          lastfmUsername?: string;
          namedQuery?: string;
          page?: number;
          size?: number;
          sort?: string[];
          spotifyEnabled?: boolean;
          username?: string;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['RaccoonUser'][];
          };
        };
      };
    };
    put?: never;
    post: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: {
        content: {
          'application/json': components['schemas']['RaccoonUser'];
        };
      };
      responses: {
        201: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['RaccoonUser'];
          };
        };
      };
    };
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/raccoon-user/count': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': number;
          };
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/raccoon-user/{id}': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          id: number;
        };
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['RaccoonUser'];
          };
        };
      };
    };
    put: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          id: number;
        };
        cookie?: never;
      };
      requestBody?: {
        content: {
          'application/json': components['schemas']['RaccoonUser'];
        };
      };
      responses: {
        201: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['RaccoonUser'];
          };
        };
      };
    };
    post?: never;
    delete: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          id: number;
        };
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        204: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/release': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          id?: number;
          musicbrainzId?: string;
          name?: string;
          namedQuery?: string;
          page?: number;
          size?: number;
          sort?: string[];
          spotifyUri?: string;
          type?: string;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['Release'][];
          };
        };
      };
    };
    put?: never;
    post: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: {
        content: {
          'application/json': components['schemas']['Release'];
        };
      };
      responses: {
        201: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['Release'];
          };
        };
      };
    };
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/release-scrape': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get?: never;
    put: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['ReleaseScrapeResponse'];
          };
        };
      };
    };
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/release/count': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': number;
          };
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/release/{id}': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          id: number;
        };
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['Release'];
          };
        };
      };
    };
    put: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          id: number;
        };
        cookie?: never;
      };
      requestBody?: {
        content: {
          'application/json': components['schemas']['Release'];
        };
      };
      responses: {
        201: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'application/json': components['schemas']['Release'];
          };
        };
      };
    };
    post?: never;
    delete: {
      parameters: {
        query?: never;
        header?: never;
        path: {
          id: number;
        };
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        204: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/scrape-taste/lastfm': {
    parameters: {
      query?: {
        email?: string;
      };
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          email?: string;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content: {
            'text/plain': components['schemas']['UserArtist'][];
          };
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/scrape-taste/spotify': {
    parameters: {
      query?: {
        email?: string;
      };
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          email?: string;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Authorized */
        401: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
        /** @description Not Allowed */
        403: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/spotify-auth-callback': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          code?: string;
          error?: string;
          state?: string;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post: {
      parameters: {
        query?: never;
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: {
        content: {
          'application/json': components['schemas']['RegisterUserRequest'];
        };
      };
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
  '/spotify-auth-callback/user-top-artists': {
    parameters: {
      query?: never;
      header?: never;
      path?: never;
      cookie?: never;
    };
    get: {
      parameters: {
        query?: {
          userId?: string;
        };
        header?: never;
        path?: never;
        cookie?: never;
      };
      requestBody?: never;
      responses: {
        /** @description OK */
        200: {
          headers: {
            [name: string]: unknown;
          };
          content?: never;
        };
      };
    };
    put?: never;
    post?: never;
    delete?: never;
    options?: never;
    head?: never;
    patch?: never;
    trace?: never;
  };
}
export type webhooks = Record<string, never>;
export interface components {
  schemas: {
    Artist: {
      /** Format: int64 */
      id?: number;
      createDate?: components['schemas']['LocalDateTime'];
      name?: string;
      lastfmUri?: string;
      spotifyUri?: string;
      musicbrainzId?: string;
      spotifyUriId?: string;
    };
    ArtistDto: {
      /** Format: int64 */
      id: number;
      name: string;
      lastfmUri?: string;
      spotifyUri?: string;
      musicbrainzId?: string;
    };
    ArtistSearchResponse: {
      /** Format: int32 */
      count?: number;
      artists?: components['schemas']['SearchResultArtistDto'][];
    };
    FollowedArtistsResponse: {
      /** Format: int32 */
      total?: number;
      rows?: components['schemas']['ArtistDto'][];
    };
    /**
     * Format: date
     * @example 2022-03-10
     */
    LocalDate: string;
    /**
     * Format: date-time
     * @example 2022-03-10T12:15:50
     */
    LocalDateTime: string;
    RaccoonUser: {
      /** Format: int64 */
      id?: number;
      email: string;
      username?: string;
      lastfmUsername?: string;
      spotifyEnabled?: boolean;
      lastNotified?: components['schemas']['LocalDate'];
      createDate?: components['schemas']['LocalDateTime'];
      modifyDate?: components['schemas']['LocalDateTime'];
      lastSpotifyScrape?: components['schemas']['LocalDateTime'];
      lastLastFmScrape?: components['schemas']['LocalDateTime'];
    };
    RegisterUserRequest: {
      email: string;
      lastfmUsername?: string;
      spotifyEnabled?: boolean;
    };
    Release: {
      /** Format: int64 */
      id?: number;
      name?: string;
      type?: string;
      spotifyUri?: string;
      musicbrainzId?: string;
      releasedOn?: components['schemas']['LocalDate'];
    };
    ReleaseScrapeResponse: {
      scrape?: components['schemas']['Scrape'];
    };
    Scrape: {
      /** Format: int64 */
      id?: number;
      createDate?: components['schemas']['LocalDateTime'];
      modifyDate?: components['schemas']['LocalDateTime'];
      completeDate?: components['schemas']['LocalDateTime'];
      /** Format: int64 */
      releasesFromSpotify?: number;
      /** Format: int64 */
      releasesFromMusicbrainz?: number;
      /** Format: int32 */
      releaseCount?: number;
      /** Format: int64 */
      usersNotified?: number;
      /** Format: int64 */
      relevantReleases?: number;
      isComplete?: boolean;
      releases?: components['schemas']['Release'][];
    };
    SearchResultArtistDto: {
      /** Format: int64 */
      id?: number;
      name: string;
      lastfmUri?: string;
      spotifyUri?: string;
      musicbrainzId?: string;
      followedByUser?: boolean;
    };
    UserArtist: {
      /** Format: float */
      weight?: number;
      hasNewRelease?: boolean;
      user?: components['schemas']['RaccoonUser'];
      artist?: components['schemas']['Artist'];
    };
  };
  responses: never;
  parameters: never;
  requestBodies: never;
  headers: never;
  pathItems: never;
}
export type $defs = Record<string, never>;
export type operations = Record<string, never>;
