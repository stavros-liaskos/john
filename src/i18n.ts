export const metaI18n = {
  title: 'Release Raccoon',
  description: 'A music release newsletter',
  url: 'https://www.releaseraccoon.online',
};
