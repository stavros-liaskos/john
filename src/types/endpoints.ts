const Paths = {
  Auth: '/api/auth/me',
  Search: `/artist/search`,
  Recommended: `/artists/recommended`,
  Follow: `/me/follow`,
  FollowedArtists: `/me/followed-artists`,
  UnfollowArtist: `/me/unfollow`,
  RaccoonUser: `/raccoon-user`,
  ScrapeLastFM: `/scrape-taste/lastfm`,
  ScrapeSpotify: `/scrape-taste/spotify`,
};

const Endpoints: Record<string, string> = {
  Auth: `${process.env.BE_BASE_URL}${Paths.Auth}`,
  Search: `${process.env.BE_BASE_URL}${Paths.Search}`,
  Recommended: `${process.env.BE_BASE_URL}${Paths.Recommended}`,
  Follow: `${process.env.BE_BASE_URL}${Paths.Follow}`,
  FollowedArtists: `${process.env.BE_BASE_URL}${Paths.FollowedArtists}`,
  UnfollowArtist: `${process.env.BE_BASE_URL}${Paths.UnfollowArtist}`,
  RaccoonUser: `${process.env.BE_BASE_URL}${Paths.RaccoonUser}`,
  ScrapeLastFM: `${process.env.BE_BASE_URL}${Paths.ScrapeLastFM}`,
  ScrapeSpotify: `${process.env.BE_BASE_URL}${Paths.ScrapeSpotify}`,
};

export default Endpoints;

export { Paths };
