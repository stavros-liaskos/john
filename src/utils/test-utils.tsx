import React from 'react';
import { render, act } from '@testing-library/react';
import ThemeProvider from '../contexts/Theme/ThemeProvider';
import { UserProvider } from '@auth0/nextjs-auth0/client';
import ArtistsListProvider from '../contexts/ArtistsList/ArtistsListProvider';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider>
      <UserProvider>
        <ArtistsListProvider>{children}</ArtistsListProvider>
      </UserProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: { [key: string]: unknown }) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const renderWithAct = async (ui: React.ReactElement, options?: { [key: string]: unknown }) => {
  const component = render(ui, { wrapper: AllTheProviders, ...options });

  const promise = Promise.resolve();
  await act(() => promise);

  return component;
};

export const beforeEachTest = () => {
  jest.resetModules();
  jest.clearAllMocks();
  jest.restoreAllMocks();
  localStorage.clear();
};
