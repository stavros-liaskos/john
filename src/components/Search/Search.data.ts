import { SearchI18n } from './Search.types';

export const searchI18n: SearchI18n = {
  label: 'Here goes the artist...',
  button: 'Search',
  searchList: {
    btnTxt: 'Follow',
    noArtists: 'No artists found',
  },
};
