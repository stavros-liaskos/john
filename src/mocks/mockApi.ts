import nock from 'nock';
import followedArtists from './fixtures/responses/followed-artists.json';
import { components } from '../types/schema';

export const nockAuth = {
  success: () => nockAPI().get(`/api/auth/me`).reply(200),
};

export const nockFollowedArtists = {
  success: (artistQuantity: number = 2) => {
    followedArtists.rows.splice(1, 2 - artistQuantity);
    nockAPI().get('/me/followed-artists').reply(200, followedArtists);
  },
};

export const nockRecommendedArtists = {
  success: () => nockAPI().get('/artists/recommended').query({ page: 0, size: 10 }).reply(200, followedArtists),
};

export const nockScrape = {
  success: () => nockAPI().get('/scrape-taste/spotify').reply(200, {}),
  fail: () => nockAPI().get('/scrape-taste/spotify').reply(200),
};

export const nockFollow = {
  success: () => nockAPI().post('/me/follow').reply(200),
};

export const nockUnfollow = {
  success: () => nockAPI().delete('/me/unfollow').reply(200),
};

export const nockSearch = {
  success: (res: components['schemas']['ArtistSearchResponse'], searchQuery: string) =>
    nockAPI().get('/artist/search').query({ pattern: searchQuery }).reply(200, res),
};

function nockAPI() {
  return nock(process.env.BE_BASE_URL!);
}
