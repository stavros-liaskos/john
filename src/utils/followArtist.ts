import { components } from '../types/schema';
import Endpoints from '../types/endpoints';

async function followArtist(artist: components['schemas']['SearchResultArtistDto'], cb: () => void) {
  const headers = new Headers({ 'Content-Type': 'application/json' });
  await fetch(Endpoints.Follow, {
    method: 'POST',
    credentials: 'include',
    headers,
    body: JSON.stringify(artist),
  })
    .then(() => {
      console.log(`${artist.name} followed successfully`);
    })
    .catch(error => {
      console.error('Error:', JSON.stringify(error));
    })
    .finally(() => {
      cb();
    });
}
export default followArtist;
