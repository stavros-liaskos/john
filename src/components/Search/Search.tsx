import { SearchProps } from './Search.types';
import React, { useState } from 'react';
import Button from '../Button/Button';
import { components } from '../../types/schema';
import FormInput from '../FormInput/FormInput';
import ArtistsList from '../ArtistsList/ArtistsList';
import followArtist from '../../utils/followArtist';
import { useArtistsListContext } from '../../contexts/ArtistsList/ArtistsListContext';
import Close from '../Icons/close';
import HandGlass from '../Icons/handGlass';

const Search: React.FunctionComponent<SearchProps> = ({ i18n }) => {
  const { getFollowedArtists } = useArtistsListContext();
  const [artistLoading, setArtistLoading] = useState<number>(0);
  const [results, setResults] = useState<components['schemas']['SearchResultArtistDto'][] | null>(null);
  const [disabled, setDisabled] = useState<boolean>(false);

  if (!i18n?.label) {
    return null;
  }
  return (
    <div className="relative flex lg:justify-center items-center flex-none h-16 md:h-20 md:border-b-2 rr-border w-full">
      <FormInput handleAction={handleSearch} i18n={i18n} actionEventTrigger={'onSubmit'}>
        {results && (
          <button onClick={() => setResults(null)}>
            <Close />
          </button>
        )}
        <Button className="btn-large md:ml-52" type="submit" disabled={disabled} loading={disabled} aria-label="search">
          <HandGlass />
        </Button>
      </FormInput>

      {results && (
        <div className="absolute px-3 bg-slate-100 dark:bg-gh-darkly border-2 rr-border top-14 md:top-16 w-full z-10">
          <ArtistsList
            i18n={i18n.searchList}
            artistsList={results}
            onButtonClick={artist => handleFollow(artist)}
            artistLoading={artistLoading}
          />
        </div>
      )}
    </div>
  );

  async function handleSearch(inputValue: string) {
    inputValue &&
      (await fetch(`/artist/search?${new URLSearchParams({ pattern: inputValue })}`, {
        method: 'GET',
        credentials: 'include',
      })
        .then(res => res.json())
        .then(result => {
          getFollowedArtists();
          return setResults(result.artists);
        })
        .catch(console.error));
  }

  async function handleFollow(artist: components['schemas']['SearchResultArtistDto']) {
    artist?.id && setArtistLoading(artist.id);
    setDisabled(true);

    const finallyCb = () => {
      artistLoading && setArtistLoading(0);
      setResults(null);
      setDisabled(false);
    };
    await followArtist(artist, finallyCb);
  }
};
Search.whyDidYouRender = true;
export default Search;
