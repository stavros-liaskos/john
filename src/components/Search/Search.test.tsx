import React from 'react';
import { fireEvent, act } from '@testing-library/react';
import Search from './Search';
import { beforeEachTest, renderWithAct } from '../../utils/test-utils';
import artistSearch from '../../mocks/fixtures/responses/artist-search.json';
import { mswAuth, mswFollowedArtists, mswSearch } from '../../mocks/mockApi';
import { components } from '../../types/schema';
import { setupServer } from 'msw/node';
import { searchI18n } from '../../i18n';

describe('Search', () => {
  const server = setupServer();

  const setup = async () => {
    const { container, getByRole } = await renderWithAct(<Search i18n={searchI18n} />);
    const searchBtn = getByRole('button');
    const input = getByRole('textbox');

    return {
      container,
      searchBtn,
      input,
    };
  };

  beforeAll(() => {
    server.listen();
    server.listen({
      onUnhandledRequest: 'error',
    });
  });
  afterEach(() => {
    jest.restoreAllMocks();
    server.resetHandlers();
  });
  afterAll(() => server.close());

  beforeEach(() => {
    beforeEachTest();
    server.use(mswAuth.success(), mswFollowedArtists.success(2));
  });

  it('renders without data without crashing', async () => {
    await renderWithAct(<Search />);
  });

  it('should have the query that the user types in the input', async () => {
    const { input } = await setup();

    fireEvent.change(input, { target: { value: 'Nels Cline' } });
    expect(input).toHaveValue('Nels Cline');
  });

  it.each<{ searchQuery: string; searchRes: components['schemas']['ArtistSearchResponse']; goal: string }>([
    { searchQuery: 'Sam Gendel', searchRes: artistSearch, goal: 'should handle the search action of the user' },
    { searchQuery: 'No match', searchRes: { artists: [], count: 0 }, goal: 'handles no search results' },
  ])('$goal', async ({ searchQuery, searchRes }) => {
    server.use(mswSearch.success(searchRes, searchQuery));
    const { container, input, searchBtn } = await setup();

    fireEvent.change(input, { target: { value: searchQuery } });
    await act(async () => {
      fireEvent.click(searchBtn);
    });

    if (searchQuery === 'No match') {
      expect(container).toHaveTextContent(searchI18n.searchList.noArtists);
    }
    expect(container).toMatchSnapshot();
  });
});
