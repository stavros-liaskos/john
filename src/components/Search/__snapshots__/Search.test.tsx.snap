// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search handles no search results 1`] = `
<div>
  <div
    class="relative flex lg:justify-center items-center flex-none h-16 md:h-20 md:border-b-2 rr-border w-full"
  >
    <form
      class="flex justify-between md:justify-between items-stretch h-10 w-full my-3"
      novalidate=""
    >
      <input
        class="mr-4 px-2 min-m-lg border-b-2 rr-border dark:bg-gh-darkly rr-text w-full"
        name="search"
        placeholder="Here goes the artist..."
        type="text"
        value="No match"
      />
      <button>
        <svg
          class="h-8 w-8 icon-colour"
          fill="none"
          role="img"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 6L6 18"
          />
          <path
            d="M6 6L18 18"
          />
        </svg>
      </button>
      <button
        class="btn flex btn-large md:ml-52"
        type="submit"
      >
        <svg
          class="h-8 w-8 icon-colour"
          fill="none"
          role="img"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </button>
    </form>
    <div
      class="absolute px-3 bg-slate-100 dark:bg-gh-darkly border-2 rr-border top-14 md:top-16 w-full z-10"
    >
      <div
        class="overflow-auto w-full flex-auto h-64"
      >
        <p
          class="rr-text flex justify-center"
        >
          No artists found
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search should handle the search action of the user 1`] = `
<div>
  <div
    class="relative flex lg:justify-center items-center flex-none h-16 md:h-20 md:border-b-2 rr-border w-full"
  >
    <form
      class="flex justify-between md:justify-between items-stretch h-10 w-full my-3"
      novalidate=""
    >
      <input
        class="mr-4 px-2 min-m-lg border-b-2 rr-border dark:bg-gh-darkly rr-text w-full"
        name="search"
        placeholder="Here goes the artist..."
        type="text"
        value="Sam Gendel"
      />
      <button
        class="btn flex btn-large md:ml-52"
        type="submit"
      >
        <svg
          class="h-8 w-8 icon-colour"
          fill="none"
          role="img"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </button>
    </form>
  </div>
</div>
`;
